src/box/gc.c

```diff
@@ -735,6 +809,238 @@ gc_consumer_name(struct gc_consumer *consumer)
+	 * Moreover, the function can be called when quorum cannot be collected,
+	 * (for example, on IPROTO_SUBSCRIBE), so we must set this flag here.
+	 */
+	txn_set_flags(txn, TXN_FORCE_ASYNC);
+	if (gc_consumer_persist(consumer) != 0) {
+		txn_abort(txn);
+		return -1;
+	}
+	if (txn_commit(txn) != 0)
+		return -1;
+	/* Unref WAL GC delay if being persisted for the first time. */
+	if (!consumer->is_persistent)
+		gc_delay_unref();
+	consumer->is_persistent = true;
+	return 0;
+}
+
+int
+gc_erase_consumer(const struct tt_uuid *uuid)
+{
+	/* Delete in-memory consumer. */
+	bool was_persistent = false;
+	struct gc_consumer *consumer = gc_consumer_by_uuid(uuid);
+	if (consumer != NULL) {
+		/* Consumer won't be asynchronously advanced anymore. */
+		was_persistent = consumer->is_persistent;
+		consumer->is_persistent = false;
+	}
+
+	/* Try to delete consumer from persistent state first. */
+	if (gc_has_persistent_consumers() &&
+	    boxk(IPROTO_DELETE, BOX_GC_CONSUMERS_ID, "[%s]",
+		 tt_uuid_str(uuid)) != 0) {
+		if (consumer != NULL)
+			consumer->is_persistent = was_persistent;
+		return -1;
+	}
+
+	if (consumer != NULL && consumer->is_orphan)
+		gc_consumer_unregister_impl(consumer);
+	return 0;
+}
+
+/**
+ * Writes all persistent gc consumers that are not marked as synchronized.
+ */
+static int
+gc_consumers_sync_persistent(void)
+{
+	int rc = 0;
+	struct txn *txn = txn_begin();
+	if (txn == NULL)
+		return -1;
+	/* No need to wait limbo, so let's set TXN_FORCE_ASYNC here. */
+	txn_set_flags(txn, TXN_FORCE_ASYNC);
+
+	struct gc_consumer *consumer;
+	rlist_foreach_entry(consumer, &gc.consumers, in_consumers) {
+		if (consumer->is_persistent &&
+		    !consumer->is_synced &&
+		    gc_consumer_persist(consumer) != 0)
+			goto fail;
+	}
+	rc = txn_commit(txn);
+	goto out;
+fail:
+	rc = -1;
+	txn_abort(txn);
+out:
+	return rc;
+}
+
+static int
+gc_sync_fiber_f(va_list ap)
+{
+	(void)ap;
+	while (!fiber_is_cancelled()) {
+		fiber_check_gc();
+		int rc = gc_consumers_sync_persistent();
+		if (rc != 0) {
+			say_error("Failed to update persistent WAL GC state");
+			diag_log();
+		}
+		fiber_sleep(TIMEOUT_INFINITY);
+	}
+	return 0;
+}
+
+/** GC consumer definition. */
+struct gc_consumer_def {
+	/** Instance UUID. */
+	struct tt_uuid uuid;
+	/** Instance vclock. */
+	struct vclock vclock;
+	/** Description of object owning this consumer. */
+	const char *object;
+};
+
+/** Mapping from tuple.opts to fields of gc_consumer_def. */
+const struct opt_def gc_consumer_def_opts_reg[] = {
+	OPT_DEF("object", OPT_STRPTR, struct gc_consumer_def, object),
+	OPT_END,
+};
+
+/**
+ * Fill gc_consumer_def with opts from the MsgPack map.
+ * Argument map can be NULL - default options are set in this case.
+ */
+static int
+gc_consumer_def_opts_decode(struct gc_consumer_def *def, const char *map,
+			    struct region *region)
+{
+	def->object = "";
+	if (map == NULL)
+		return 0;
+	return opts_decode(def, gc_consumer_def_opts_reg, &map, region);
+}
```

Comment:
Nit: Better inline this function - it's a few lines and used only once.

Reply:
Done.